### PR TITLE
Fixes #28528

### LINKED ISSUE #28528: read_link cannot open some files reported as symbolic links on windows
rustc on Windows 7 32 bit:

```
C:\Users\andrew\walkdir\symlinks>rustc --version
rustc 1.5.0-nightly (6e5a32547 2015-09-19)
```

Test program:

```rust
use std::fs;

fn main() {
    let md = fs::symlink_metadata("is_a_symlink").unwrap();
    println!("is symlink? {}", md.file_type().is_symlink());
    match fs::read_link("is_a_symlink") {
        Ok(p) => println!("{}", p.display()),
        Err(err) => println!("{:?}", err),
    }
}
```

Create a symlink on Windows at `is_a_symlink` pointing to `not_a_symlink` and
run above program (**working as intended**):

```
C:\Users\andrew\walkdir\symlinks>mklink is_a_symlink C:\Users\andrew\walkdir\symlinks\not_a_symlink
symbolic link created for is_a_symlink <<===>> C:\Users\andrew\walkdir\symlinks\not_a_symlink

C:\Users\andrew\walkdir\symlinks>dir
09/19/2015  11:40 PM    <DIR>          .
09/19/2015  11:40 PM    <DIR>          ..
09/19/2015  11:40 PM    <SYMLINK>      is_a_symlink [C:\Users\andrew\walkdir\symlinks\not_a_symlink]
09/19/2015  11:32 PM                 0 not_a_symlink
09/19/2015  11:36 PM         2,118,989 win_symlink_test.exe
09/19/2015  11:36 PM               261 win_symlink_test.rs

C:\Users\andrew\walkdir\symlinks>win_symlink_test.exe
is symlink? true
\??\C:\Users\andrew\walkdir\symlinks\not_a_symlink
```

This correctly identifies the file as a symlink and resolves it correctly.

If we instead creation a junction and follow the same steps, we can observe an
inconsistency between `file_type().is_symlink()` and the behavior of
`fs::read_link`:

```
C:\Users\andrew\walkdir\symlinks>mklink /J is_a_symlink C:\Users\andrew\walkdir\symlinks\not_a_symlink
Junction created for is_a_symlink <<===>> C:\Users\andrew\walkdir\symlinks\not_a_symlink

C:\Users\andrew\walkdir\symlinks>dir
09/19/2015  11:46 PM    <DIR>          .
09/19/2015  11:46 PM    <DIR>          ..
09/19/2015  11:46 PM    <JUNCTION>     is_a_symlink [C:\Users\andrew\walkdir\symlinks\not_a_symlink]
09/19/2015  11:32 PM                 0 not_a_symlink
09/19/2015  11:44 PM         2,118,946 win_symlink_test.exe
09/19/2015  11:44 PM               286 win_symlink_test.rs

C:\Users\andrew\walkdir\symlinks>win_symlink_test.exe
is symlink? true
Error { repr: Custom(Custom { kind: Other, error: StringError("not a symlink") }) }
```

(The key difference is the use of `mklink /J` to create a junction.)

The output here shows that it is correctly detected as a symlink, but
`read_link` chokes on it.

I looked briefly at the implementation and I haven't the slightest clue what's
wrong (I'm not familiar with Windows).

### PR DESCRIPTION
**Pull request by pitdicker**

### CODE REVIEW

File: src/libstd/sys/windows/fs.rs [hunk 1]
```diff
@@ -371,19 +367,34 @@ impl File {
...
-            let subst_off = (*info).SubstituteNameOffset / 2;
+            let (path_buffer, subst_off, subst_len, relative) = match buf.ReparseTag {
+                c::IO_REPARSE_TAG_SYMLINK => {
+                        let info: *const c::SYMBOLIC_LINK_REPARSE_BUFFER =
+                            &buf.rest as *const _ as *const _;
+                        (&(*info).PathBuffer as *const _ as *const u16,
+                         (*info).SubstituteNameOffset / 2,
+                         (*info).SubstituteNameLength / 2,
+                         (*info).Flags & c::SYMLINK_FLAG_RELATIVE != 0)
+                    },
```
Comment (alexcrichton) [inline src/libstd/sys/windows/fs.rs]: Stylistically we don't indent over the match body an extra indent

**Comment by alexcrichton** [comment]
Thanks @pitdicker! Just a few stylistic nits, but otherwise looks good to me.

Does all metadata on Windows go through `File::metadata` now? I think that wasn't true at some point but it now looks to be the case? We could change the layout of `fs::Metadata` if that's the case to just be exactly that return value.

**Comment by alexcrichton** [comment]
Ah and also can you expand the commit message to explain what's going on as well?

**Comment by pitdicker** [comment]
I am not completely sure what you mean about metadata. There are now only two points that call into WinAPI for metadata: `File::file_attr` and in `DirEntry`. But what could be changed to be some return variable?

**Comment by alexcrichton** [comment]
Ah yeah I guess we still have two entry points, perhaps we can refactor later!

@bors: r+ 6403f913dd102ed61626c9c1c6607731915957f6

Thanks!

**Comment by pitdicker** [comment]
Thanks for the review!

An other question. Many tests and private functions now have Unix names like `mkdir` and `unlink`, instead of `create_dir` and `delete`. Would it make sense to rename them?
It would make it a very little bit easier to find the function of a failing test, and to lift code up/down from `sys::*::fs` to `fs`.

**Comment by pitdicker** [comment]
I don't think there is much refactoring we can do for metadata, but will think about it.

**Comment by pitdicker** [comment]
This looks like a timeout, the other Windows builds succeeded

**Pull request merged**